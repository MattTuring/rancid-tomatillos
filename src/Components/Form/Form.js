import React from 'react'
import { withRouter } from 'react-router-dom';
import { connect } from 'react-redux';
import { addUserState, addRatings, changeLoading } from '../../actions';
import { postUser, getRatings } from '../../fetchcalls';
import PropTypes from 'prop-types';

export class Form extends React.Component {
  constructor() {
    super()
    this.state = {
      id:'',
      username: '',
      passwordLength: '',
      error: null,
      loggedIn: false,
      password: null
    }
  }

  login = () => {
    if (this.state.username === '' || this.state.passwordLength > 0) {
      this.setState({ error: 'THE USERNAME OR PASSWORD IS INCORECT' })
    }
    this.props.changeLoading();
    postUser('https://rancid-tomatillos.herokuapp.com/api/v1/login', {email: this.state.username, password: this.state.password})
      .then(data => {
        this.setState({ id: data.user.id, loggedIn: true },() =>this.props.addUserState({
          id: this.state.id,
          username: this.state.username,
          ratings: [],
          loggedIn: this.state.loggedIn,
          password: null
        }))
        this.props.history.push(`/users/${data.user.id}`)
        getRatings(this.state.id)
          .then(data => {
            this.props.addRatings(data);
            this.props.changeLoading();
          })
      })
      .catch(error => console.log(error))
  }

  render() {
    return (
      <article className='login'>
        <label htmlFor='username'>USERNAME</label>
        <input name='username' id='username' type='text' placeholder='USERNAME' onChange={(event) => this.setState({username:event.target.value})}/>
        <label htmlFor='password'>PASSWORD</label>
        <input name='password' id='password' type='password' placeholder='PASSWORD' onChange={(event) => {this.setState({password:event.target.value}); this.setState({passwordLength:event.target.value.length});}}/>
        <button className='login-button' onClick={this.login}>LOG IN</button>
        {this.state.error}
      </article>
    )
  }
}

export const mapDispatchToProps = dispatch => ({
  addUserState: userinfo => dispatch(addUserState(userinfo)),
  addRatings: (ratings) => dispatch(addRatings( ratings )),
  changeLoading: () => dispatch(changeLoading())
})

export default connect(null, mapDispatchToProps)(withRouter(Form))

Form.propTypes = {
  id: PropTypes.number,
  username: PropTypes.string,
  passwordLength: PropTypes.number,
  error: PropTypes.string,
  loggedIn: PropTypes.bool,
  password: PropTypes.string,
}
