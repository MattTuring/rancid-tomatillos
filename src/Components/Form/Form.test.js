import React from 'react';
import { shallow } from 'enzyme';
import { Form }  from './Form';

describe('Form', () => {
    let wrapper

    beforeEach(() => {
        wrapper = shallow(<Form />)
    });

    it('should match the snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    });


    it('should return error if login validation fails', () => {
        const originalState = {
            id: '',
            username: '',
            password: '',
            error: null,
            loggedIn: false
        };

        wrapper.setState(originalState)

        wrapper.instance().login()

        expect(wrapper.state()).toEqual({
            error: "THE USERNAME OR PASSWORD IS INCORECT",
            id: "",
            loggedIn: false,
            password: "",
            username: ""
        });
    });

    it('should update state when login is called with input values', () => {
        const populatedState = {
            id: '1',
            username: 'foster',
            password: 'abc',
            error: null,
            loggedIn: false
        };

        const expected =  {
            id: '1',
            username: 'foster',
            password: 'abc',
            error: null,
            loggedIn: false
        };

        wrapper.setState(populatedState)
        // mock http request
        wrapper.instance().login()

        expect(wrapper.state()).toEqual(expected);
    });
});
