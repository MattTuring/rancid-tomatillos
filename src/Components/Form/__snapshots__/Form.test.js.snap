// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form should match the snapshot 1`] = `
<article
  className="login"
>
  <label
    htmlFor="username"
  >
    USERNAME
  </label>
  <input
    id="username"
    name="username"
    onChange={[Function]}
    placeholder="USERNAME"
    type="text"
  />
  <label
    htmlFor="password"
  >
    PASSWORD
  </label>
  <input
    id="password"
    name="password"
    onChange={[Function]}
    placeholder="PASSWORD"
    type="password"
  />
  <button
    className="login-button"
    onClick={[Function]}
  >
    LOG IN
  </button>
</article>
`;
